mvebu: provide U-boot and ATF for Methode eDPU

robimarko opened the pull request:
Backport U-boot support for Methode eDPU as well as provide ATF builds for it.
U-boot patches can be dropped once 2022.10 is released and we update to it.

This makes it easy for OpenWrt users to update the included U-boot+ATF combo.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
uboot-mvebu: add Methode eDPU support

Add support for building for Methode eDPU board, no patches are needed
as board has been upstreamed and is part of the 2022.10-rc releases.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
arm-trusted-firmware-mvebu: add Methode eDPU support

Provide ATF support for Methode eDPU as well, this makes it easy for
OpenWrt users to update the included U-boot+ATF combo.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
@BKPepe You are most active on ATF and U-boot MVEBU matters

BKPepe commented:
> U-boot patches can be dropped once 2022.10 is released and we update to it.

I think wouldn't it be easier for you to use 2022.10-rc5? It will be even better for me as well. :-) Because we (in Turris) are using this version already on Turris Omnia routers with additional backports, which we need and are useful for mvebu routers as well. But it is a question if we want to have rc version in the master.

robimarko commented:
It would be easier in the sense that nothing would be backported, I dont have anything against RC releases myself

dhewg commented:
I'd be interested to test the latest rc on espressobin as well, and since this is a per arch package this should be okay (famous last words)

robimarko commented:
I am fine with using RC-s, it's not like final releases are something special in U-boot or receive additional testing.

BKPepe commented on package/boot/arm-trusted-firmware-mvebu/Makefile:
```diff
@@ -105,7 +115,8 @@ TFA_TARGETS:= \
 	espressobin-v7-2gb \
-	udpu
+	udpu \
+  edpu
```
It seems this has wrong indentation.

robimarko replied:
Thanks for spotting this, its a constant battle since there is a mix of tab and space indentation everywhere

BKPepe replied:
Yeah. It also depends which text editor(s) you are using, etc. I am quite familiar with this. No worries!

robimarko replied:
I am using Sublime which as a rule if indents are mixed defaults to spaces, personally I just wish tabs were used everywhere in the buildroot

hauke commented:
Uboot 2022.10 will be released soon:
`Release "v2022.10" is scheduled in 11 days — on 03 October 2022. `
https://www.denx.de/wiki/U-Boot/ReleaseCycle

BKPepe commented:
Oh, already? The time flies a little bit faster than I thought. Thanks!

For the new batches, we used [rc4-rc5 ](https://gitlab.nic.cz/turris/os/build/-/commit/959e13e1a142a6e294dc90598ff77c7e393a3a20)versions with some more fixes. That's why I suggested using it, because we are testing OpenWrt U-boot for mvebu (Turris Omnia, Turris MOX) as we copied it from the staging image dir to our own package, so users can install it if they want to provide them more friendly way. This helps us not to compile U-boot twice.

BKPepe reviewed (approved):
Looks good! Thanks!
- Patches were backported from upcoming U-boot version, thus they have all the necessary things which should have (commit subject, commit description, etc.).

Unfortunately, I don't have this device to test it on, but that should not this block this PR:

robimarko commented:
@hauke So, should I wait until its released or this is fine?

dhewg commented:
We can probably catch more regression from v2022.07 to the next -rcX than from -rcX to final.

So how about updating this to -rc, so we can all test the very same thing, including any additional patches you require?
If the release is out before this gets merged, you can just update it again.

And I suspect an uboot v2022.07 issue behind #10583, so @arrmo might be interested in testing this too without working around it with governor hacks.

arrmo commented:
Willing to give this a try! Pull in your changes, rebuild, and re-flash U-Boot?

Thanks!

robimarko committed:
uboot-mvebu: update to 2022.10

Update mvebu U-boot to 2022.10 to avoid backporting patches in order
to support Methode eDPU.

It also allows dropping existing patches as they are all backports.

Tested-by: Andre Heider <a.heider@gmail.com> # espressobin-v3-v5-1gb-2cs
Tested-by: Russell Morris <github@rkmorris.us> # espressobin-v3-v5-1gb-1cs
Tested-by: Josef Schlehofer <pepe.schlehofer@gmail.com> [Turris Omnia]
Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
@dhewg Ok, lets go down the route of updating U-boot to RC5

dhewg commented:
Thanks, built and flashed! Looks good at first glance on espressobin, but I'll wait at least a week until judging if it's stable.

@arrmo exactly, basically just a `curl -L "https://github.com/openwrt/openwrt/pull/10776.patch"|git am`, build it and flash it. On reboot you should see `U-Boot 2022.10-rc5-OpenWrt-*` on serial. Please test drive that without your governor hacks.

arrmo commented:
> curl -L "https://github.com/openwrt/openwrt/pull/10776.patch"|git am

Hmmm, tried that, got,
```
Applying: uboot-mvebu: update to 2022.10-rc5
Applying: uboot-mvebu: add Methode eDPU support
Applying: arm-trusted-firmware-mvebu: add Methode eDPU support
error: patch failed: package/boot/arm-trusted-firmware-mvebu/Makefile:105
error: package/boot/arm-trusted-firmware-mvebu/Makefile: patch does not apply
Patch failed at 0003 arm-trusted-firmware-mvebu: add Methode eDPU support
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```
Thoughts?

Thanks!

dhewg commented:
It should work on current master, in fact I copied that cmdline from my shell history :)
So you probably need to rebase your tree or you have your own patches there?
But first get rid of the half-applied state with `git am --abort`.

Oh, while here, to clarify: what I mean by `flash it` before was just updating the bootloader of course, so `bubt flash-image.bin spi tftp` or similar from the u-boot prompt. To properly test this, don't update the whole of OpenWrt just yet. Only change one piece to test just that.

dhewg commented:
Also, since it fails at arm-trusted-firmware-mvebu for you: did you change CLOCKPRESET by chance? If so, just don't, it's not worth it. It depends on your very board of course, but that's one major root of instability

arrmo commented:
> Oh, while here, to clarify: what I mean by `flash it` before was just updating the bootloader of course, so `bubt flash-image.bin spi tftp` or similar from the u-boot prompt. To properly test this, don't update the whole of OpenWrt just yet. Only change one piece to test just that.

Yep, understood - and agreed!



> It should work on current master, in fact I copied that cmdline from my shell history :)

OK, I was trying on top of v22.03.0, that may be the issue. Let me try on master. But then => OK to flash, and use with v22.03.0 (OpenWrt)?

> did you change CLOCKPRESET by chance?

Nope, not changed.

Thanks!

arrmo commented:
OK, checked out master - then the patch applies cleanly! But ... the build fails 😞. Output below.
```
make -C image compile TARGET_BUILD=
make[5]: Entering directory '/mnt/ProgSSD/openwrt/target/linux/mvebu/image'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/tmp/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/tmp/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz.gz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz.gz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/json_info_files/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz.json'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/json_info_files/openwrt-mvebu-cortexa53-methode_udpu-firmware.tgz.json'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/tmp/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/tmp/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz.gz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz.gz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/bin/targets/mvebu/cortexa53/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz'
Makefile:197: warning: overriding recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/json_info_files/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz.json'
Makefile:197: warning: ignoring old recipe for target '/mnt/ProgSSD/openwrt/build_dir/target-aarch64_cortex-a53_musl/json_info_files/openwrt-mvebu-cortexa53-methode_edpu-firmware.tgz.json'
make[5]: Nothing to be done for 'compile'.
make[5]: Leaving directory '/mnt/ProgSSD/openwrt/target/linux/mvebu/image'
make[4]: Leaving directory '/mnt/ProgSSD/openwrt/target/linux/mvebu'
make[3]: Leaving directory '/mnt/ProgSSD/openwrt/target/linux'
time: target/linux/compile#10.58#3.09#13.53
make[2]: Leaving directory '/mnt/ProgSSD/openwrt'
export MAKEFLAGS= ;make -w -r buildinfo
make[2]: Entering directory '/mnt/ProgSSD/openwrt'
 make[3] diffconfig buildversion feedsversion
 make[4] scripts/config/conf
make[2]: Leaving directory '/mnt/ProgSSD/openwrt'
make[2]: Entering directory '/mnt/ProgSSD/openwrt'
rm -rf /mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/root-mvebu
make[2]: Leaving directory '/mnt/ProgSSD/openwrt'
make[2]: Entering directory '/mnt/ProgSSD/openwrt'
make[3]: Entering directory '/mnt/ProgSSD/openwrt/package/libs/libjson-c'
make[3]: Leaving directory '/mnt/ProgSSD/openwrt/package/libs/libjson-c'
time: package/libs/libjson-c/host-compile#0.18#0.07#0.22
make[3]: Entering directory '/mnt/ProgSSD/openwrt/package/libs/libubox'
MAKEFLAGS="" /mnt/ProgSSD/openwrt/staging_dir/host/bin/ninja -v -j1 -C /mnt/ProgSSD/openwrt/build_dir/hostpkg/libubox-2022-05-15-d2223ef9
ninja: Entering directory `/mnt/ProgSSD/openwrt/build_dir/hostpkg/libubox-2022-05-15-d2223ef9'
[1/3] : && /mnt/ProgSSD/openwrt/staging_dir/host/bin/gcc -O2 -I/mnt/ProgSSD/openwrt/staging_dir/host/include -I/mnt/ProgSSD/openwrt/staging_dir/hostpkg/include -I/mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/include -DNDEBUG -L/mnt/ProgSSD/openwrt/staging_dir/host/lib -L/mnt/ProgSSD/openwrt/staging_dir/hostpkg/lib -L/mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/lib   -rdynamic CMakeFiles/jshn.dir/jshn.c.o -o jshn  -Wl,-rpath,/mnt/ProgSSD/openwrt/build_dir/hostpkg/libubox-2022-05-15-d2223ef9:  libblobmsg_json.so  -Wl,-Bstatic  -ljson-c  -Wl,-Bdynamic  libubox.so && :
FAILED: jshn
: && /mnt/ProgSSD/openwrt/staging_dir/host/bin/gcc -O2 -I/mnt/ProgSSD/openwrt/staging_dir/host/include -I/mnt/ProgSSD/openwrt/staging_dir/hostpkg/include -I/mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/include -DNDEBUG -L/mnt/ProgSSD/openwrt/staging_dir/host/lib -L/mnt/ProgSSD/openwrt/staging_dir/hostpkg/lib -L/mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/lib   -rdynamic CMakeFiles/jshn.dir/jshn.c.o -o jshn  -Wl,-rpath,/mnt/ProgSSD/openwrt/build_dir/hostpkg/libubox-2022-05-15-d2223ef9:  libblobmsg_json.so  -Wl,-Bstatic  -ljson-c  -Wl,-Bdynamic  libubox.so && :
/usr/bin/ld: libubox.so: undefined reference to `.LASF706'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
make[3]: *** [Makefile:114: /mnt/ProgSSD/openwrt/build_dir/hostpkg/libubox-2022-05-15-d2223ef9/.built] Error 1
make[3]: Leaving directory '/mnt/ProgSSD/openwrt/package/libs/libubox'
time: package/libs/libubox/host-compile#0.21#0.07#0.25
    ERROR: package/libs/libubox [host] failed to build.
make[2]: *** [package/Makefile:114: package/libs/libubox/host/compile] Error 1
make[2]: Leaving directory '/mnt/ProgSSD/openwrt'
make[1]: *** [package/Makefile:110: /mnt/ProgSSD/openwrt/staging_dir/target-aarch64_cortex-a53_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory '/mnt/ProgSSD/openwrt'
make: *** [/mnt/ProgSSD/openwrt/include/toplevel.mk:231: world] Error 2
```
Thoughts? BTW, I also tried a `make dirclean`, get everything back to a clean state, but no help.

Thanks!

dhewg commented:
Hm, that doesn't look familiar, maybe a temporary regression on master?
Does just building the firmware suceeds? `make V=cs package/arm-trusted-firmware-mvebu/{clean,compile}`

arrmo commented:
> Hm, that doesn't look familiar, maybe a temporary regression on master? Does just building the firmware suceeds? `make V=cs package/arm-trusted-firmware-mvebu/{clean,compile}`

Yes, it does (build) - thanks! Like you say, perhaps a temporary regression on master.

arrmo commented:
Rebooted, flashed, and ... didn't make it 5 min 😞. Rebooted while set for `ondemand`. Trying `performance` again.

Thanks.

dhewg commented:
Aww, oh well, it was worth a try...

robimarko commented:
That's not gonna be related to U-boot, its probably one of the wonderful silicon erratas related to AVS

BKPepe commented:
I agree with @robimarko , we triggered somehow this issue with recent LTS kernel 5.15 on Turris MOX (same SoC) and we don't know, why it happens. :shrug: Marvell is not responding to us. It does not even work to decrease frequency, fiddling around voltage, etc. :-/

arrmo commented:
Wasn't aware of this, and it makes sense to me (kernel related). Not quite sure what to do here for now ... I guess just leave `performance` as the governor.

dhewg commented:
Yeah, with that test result it indeed sounds unlikely this is an uboot issue.

I do have random reboots with v2022.07 though. v2022.04 didn't have this issue, and reverting to that appears to fix it (I say appears because sometimes it can days until it happens). v2022.10-rc5 looks good so far too. This is why I wanted to check if @arrmo was affected by the same issue.

BKPepe commented:
> I do have random reboots with v2022.07 though. 

I'm thinking, do you have some logs from these random reboots? Not sure if it is even worth asking you for that since the previous and the next version of U-boot is working well for you.

> Well, 1.2GHz models have no scaling and are broken for over a year now.

To be honest, SoC A3720 is so fucking broken, and it is not just about scaling. I am fully aware of all A3720 issues, which we have, but it does not affect only us (I mean Turris devices based on that SoC) but also everyone who is using A3720 SoC. I am already getting tired of it. I think we need to be more open about those issues which we have and Marvell does not want to help us or respond to us. Imagine the situation when we have a Turris MOX router, which can not [perform reboot reliably](https://gitlab.nic.cz/turris/os/build/-/issues/82), but recently it started to show up that random reboots are happening with newer OpenWrt versions with LTS kernel 5.15, and that's not related to U-boot.

I will ask, if we can publish the internal issue, which we have for all unresolved issues for A3720.

But we are slightly getting off-topic here on this.

dhewg commented:
> I'm thinking, do you have some logs from these random reboots? Not sure if it is even worth asking you for that since the previous and the next version of U-boot is working well for you.

I tried, I had picocom running in tmux for weeks. There's nothing, it doesn't seem to crash, or at least there's no output whatsoever. It just reboots, I only ever saw `TIM` and so on as with a normal power cycle...

arrmo commented:
> I agree with @robimarko , we triggered somehow this issue with recent LTS kernel 5.15 on Turris MOX (same SoC) and we don't know, why it happens. 🤷 Marvell is not responding to us. It does not even work to decrease frequency, fiddling around voltage, etc. :-/

For this device, does moving from `ondemand` to `performance` governor also work around the issue? That seems to be a fix for me ... and not a big deal, low power device, and not getting hot at all.

I'm happy to try to debug this, but when it happens => just a reboot, no info at all. Even the console - nothing. Is there a way to enable more detailed logging, for debug? And ... to a file, so I don't need serial to retrieve the info after a reboot?

Thanks!

robimarko commented:
Bumped to the final 2022.10 as it was released yesterday, it allows dropping recent backports as well.

arrmo commented:
> Bumped to the final 2022.10 as it was released yesterday, it allows dropping recent backports as well.

Nice! Will try to build it here ... do I just cherry-pick from this commit?
https://github.com/openwrt/openwrt/commit/9238dbd5a80ee85d0493dbbfa897affcedd25dd4

Thanks!

robimarko commented:
If you only care about 2022.10 then just pick https://github.com/openwrt/openwrt/pull/10776/commits/2f021e2b1089c31736d98f61d99ef65697b96828

robimarko commented:
Yeah, its not merged, you need to download it as patch

arrmo commented:
OK, let me figure that out - I'm the definition of git-idiot 😜 

I also want to know how to undo it later, when this is merged.

Thanks!

robimarko commented:
Just `wget https://github.com/openwrt/openwrt/commit/2f021e2b1089c31736d98f61d99ef65697b96828.patch` and then `git am 2f021e2b1089c31736d98f61d99ef65697b96828.patch`

Later you can just drop the commit using `git rebase`

arrmo commented:
That makes sense, thanks! Tried it ...
```
Applying: uboot-mvebu: update to 2022.10
error: patch failed: package/boot/uboot-mvebu/Makefile:8
error: package/boot/uboot-mvebu/Makefile: patch does not apply
Patch failed at 0001 uboot-mvebu: update to 2022.10
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

robimarko commented:
Make sure that you are up to date with master

arrmo commented:
I "think" I am ... LOL,
```
git fetch upstream
git merge upstream/master master
```

Results in `Already up to date.`

Thanks!

robimarko commented:
Well, just tried fetching the patch and it applies cleanly on OpenWrt master a5265497a4f6da158e95d6a450cb2cb6dc085cab, so I dont know whats the issue

arrmo commented:
Found it! 😄. I had your earlier patch applied - and `git rebase` doesn't remove it => had to do a hard git reset.

Building now ...

Thanks!

robimarko commented:
@BKPepe Any chance you gave 2022.10 a spin?

BKPepe reviewed (approved):
Package: ``uboot-mvebu 2022.10`` (final): tested on Turris Omnia

If you want, you can add my ``Tested-by``. Would be nice to have this backported to OpenWrt 22.03 release. At least the U-boot update.

arrmo commented:
FYI, seems to be running fine here as well - I just need the governor update (boot.scr), but that seems to be kernel related, not U-Boot.

Thanks!

dhewg commented:
-rc5 was stable for ~10 days, and I've updated to the final uboot ~4 days ago - no issues so far :)

So if you care, for the u-boot commit:
`Tested-by: Andre Heider <a.heider@gmail.com> # espressobin-v3-v5-1gb-2cs`

BKPepe commented:
Guys, can we have this merged? It is approved and tested by several people including me.

cc: @Ansuel and ping @hauke :)

arrmo commented:
Running great here as well, no issues at all.

Thanks!

Ansuel commented:
@BKPepe i wonder where we should put the tested by tag? Also can you provide also yours?

@arrmo can you also provide tested by tag?


Currently i have only this

```
Tested-by: Andre Heider <a.heider@gmail.com> # espressobin-v3-v5-1gb-2cs
```

arrmo commented:
Sorry, not quite sure what you're after - I pulled in the latest changes noted above.

Thanks!

Ansuel commented:
@arrmo if you want you can leave a tag that """certify"" that you tested this patch and that it does work for your device... having many of them makes the patch more valuable as there are actual people that tested it aside from the committer... so do you want to provide a tag like

`Tested-by: Andre Heider <a.heider@gmail.com> # espressobin-v3-v5-1gb-2cs` ?

arrmo commented:
Sure, NP!
`Tested-by: Russell Morris <github@rkmorris.us> # espressobin-v3-v5-1gb-1cs`

Thanks!

BKPepe commented:
> i wonder where we should put the tested by tag?

Don't reinvent wheel and take a look at Git history, how it was done in the past. For example: https://github.com/openwrt/openwrt/commit/a2122b518e0e7e88e29753f9343e8d04e00c5ac7

> Also can you provide also yours?

Sure, I already approve it here. Tested on Turris Omnia as previous U-boot update. It can be taken from history as well.

We should not block this for more days than it is necessary, though.

Ansuel commented:
@robimarko ok can you rebase on top of master so i can merge this and flag this as merged?

the 3 tested by tag to add

```
Tested-by: Andre Heider <a.heider@gmail.com> # espressobin-v3-v5-1gb-2cs
Tested-by: Russell Morris <github@rkmorris.us> # espressobin-v3-v5-1gb-1cs
Tested-by: Josef Schlehofer <pepe.schlehofer@gmail.com> [Turris Omnia]
```

robimarko commented:
@Ansuel done
